Regression: Unexpected dereference of returned struct pointer

I've observed a regression caused by https://github.com/CloudyKit/jet/commit/91ac9cb0201cc9d6b4130277c2f7e4b3685850f6 (#97)

### Summary
For this template:
```handlebars
{{apple := GetAppleByName("honeycrisp")}}
{{TellFlavor(apple)}}
```

With this code:
```go
func main() {

	// ...

	apples := map[string]*Apple{
		"honeycrisp": {
			Flavor: "crisp",
		},
		"red-delicious": {
			Flavor: "poor",
		},
		"granny-smith": {
			Flavor: "tart",
		},
	}

	// ...

	vars.SetFunc("GetAppleByName", func(a jet.Arguments) reflect.Value {
		name := a.Get(0).String()
		return reflect.ValueOf(apples[name])
	})

	vars.SetFunc("TellFlavor", func(a jet.Arguments) reflect.Value {
		apple := a.Get(0).Interface().(*Apple)
		flav := apple.GetFlavor()
		return reflect.ValueOf(flav)
	})

	// ...
}

type Apple struct {
	Flavor string
}

func (a *Apple) GetFlavor() string {
	return a.Flavor
}
```

The expected output would be `crisp` and this worked fine with v2.1.2

After https://github.com/CloudyKit/jet/commit/91ac9cb0201cc9d6b4130277c2f7e4b3685850f6 the output is:
```
panic: interface conversion: interface {} is main.Apple, not *main.Apple
// stack trace elided
```

### Example Project
https://github.com/tooolbox/jet-example has two branches that cover this:
- v2.1.2-pointers (working)
- v2.1.3-91ac9cb (broken)

### Solution
It seems as though `{{apple := GetAppleByName("honeycrisp")}}` is returning a `*main.Apple` but the new code is dereferencing that pointer which is unexpected.  I can see where this might be handy for a `*string` or `*int` but for a `*struct{}` I can't think why this would be the desired behavior.

--- a/eval.go
+++ b/eval.go
@@ -55,7 +55,7 @@ func (renderer RendererFunc) Render(r *Runtime) {
 // Ranger a value implementing a ranger interface is able to iterate on his value
 // and can be used directly in a range statement
 type Ranger interface {
-	Range() (interface{}, interface{}, bool)
+	Range() (reflect.Value, reflect.Value, bool)
 }
 
 type escapeeWriter struct {
@@ -221,6 +221,9 @@ func (state *Runtime) Resolve(name string) reflect.Value {
 }
 
 func (st *Runtime) recover(err *error) {
+	// reset state scope and context just to be safe (they might not be cleared properly if there was a panic while using the state)
+	st.scope = &scope{}
+	st.context = reflect.Value{}
 	pool_State.Put(st)
 	if recovered := recover(); recovered != nil {
 		var is bool
@@ -457,21 +460,21 @@ func (st *Runtime) executeList(list *ListNode) {
 					if isSet {
 						if isLet {
 							if isKeyVal {
-								st.variables[node.Set.Left[0].String()] = reflect.ValueOf(indexValue)
-								st.variables[node.Set.Left[1].String()] = reflect.ValueOf(rangeValue)
+								st.variables[node.Set.Left[0].String()] = indexValue
+								st.variables[node.Set.Left[1].String()] = rangeValue
 							} else {
-								st.variables[node.Set.Left[0].String()] = reflect.ValueOf(rangeValue)
+								st.variables[node.Set.Left[0].String()] = rangeValue
 							}
 						} else {
 							if isKeyVal {
-								st.executeSet(node.Set.Left[0], reflect.ValueOf(indexValue))
-								st.executeSet(node.Set.Left[1], reflect.ValueOf(rangeValue))
+								st.executeSet(node.Set.Left[0], indexValue)
+								st.executeSet(node.Set.Left[1], rangeValue)
 							} else {
-								st.executeSet(node.Set.Left[0], reflect.ValueOf(rangeValue))
+								st.executeSet(node.Set.Left[0], rangeValue)
 							}
 						}
 					} else {
-						st.context = reflect.ValueOf(rangeValue)
+						st.context = rangeValue
 					}
 					st.executeList(node.List)
 					indexValue, rangeValue, end = ranger.Range()
@@ -605,7 +608,12 @@ func (st *Runtime) evalPrimaryExpressionGroup(node Expression) reflect.Value {
 			return baseExpression.MapIndex(indexExpression)
 		case reflect.Array, reflect.String, reflect.Slice:
 			if canNumber(indexType.Kind()) {
-				return baseExpression.Index(int(castInt64(indexExpression)))
+				index := int(castInt64(indexExpression))
+				if 0 <= index && index < baseExpression.Len() {
+					return baseExpression.Index(index)
+				} else {
+					node.errorf("%s index out of range (index: %d, len: %d)", baseExpression.Kind().String(), index, baseExpression.Len())
+				}
 			} else {
 				node.errorf("non numeric value in index expression kind %s", baseExpression.Kind().String())
 			}
@@ -1091,14 +1099,6 @@ func (st *Runtime) evalBaseExpressionGroup(node Node) reflect.Value {
 			node.errorf("identifier %q is not available in the current scope %v", node, st.variables)
 		}
 
-		// limit the number of pointers to follow
-		for dereferenceLimit := 2; resolved.Kind() == reflect.Ptr && dereferenceLimit >= 0; dereferenceLimit-- {
-			if resolved.IsNil() {
-				return reflect.ValueOf("")
-			}
-			resolved = reflect.Indirect(resolved)
-		}
-
 		return resolved
 	case NodeField:
 		node := node.(*FieldNode)
@@ -1480,7 +1480,6 @@ func getFieldOrMethodValue(key string, v reflect.Value) reflect.Value {
 }
 
 func getValue(key string, v reflect.Value) reflect.Value {
-
 	if !v.IsValid() {
 		return reflect.Value{}
 	}
@@ -1604,11 +1603,11 @@ type sliceRanger struct {
 	i   int
 }
 
-func (s *sliceRanger) Range() (index, value interface{}, end bool) {
+func (s *sliceRanger) Range() (index, value reflect.Value, end bool) {
 	s.i++
-	index = s.i
+	index = reflect.ValueOf(s.i)
 	if s.i < s.len {
-		value = s.v.Index(s.i).Interface()
+		value = s.v.Index(s.i)
 		return
 	}
 	pool_sliceRanger.Put(s)
@@ -1620,9 +1619,8 @@ type chanRanger struct {
 	v reflect.Value
 }
 
-func (s *chanRanger) Range() (_, value interface{}, end bool) {
-	_value, end := s.v.Recv()
-	value = _value.Interface()
+func (s *chanRanger) Range() (_, value reflect.Value, end bool) {
+	value, end = s.v.Recv()
 	if end {
 		pool_chanRanger.Put(s)
 	}
@@ -1636,11 +1634,10 @@ type mapRanger struct {
 	i    int
 }
 
-func (s *mapRanger) Range() (index, value interface{}, end bool) {
+func (s *mapRanger) Range() (index, value reflect.Value, end bool) {
 	if s.i < s.len {
-		_index := s.keys[s.i]
-		index = _index.Interface()
-		value = s.v.MapIndex(_index).Interface()
+		index = s.keys[s.i]
+		value = s.v.MapIndex(index)
 		s.i++
 		return
 	}

--- a/eval_test.go
+++ b/eval_test.go
@@ -373,11 +373,19 @@ func TestEvalSliceExpression(t *testing.T) {
 	RunJetTest(t, nil, []string{"111"}, "SliceExpressionSlice_IfLen", `{{if len(.) > 0}}{{.[0]}}{{end}}`, `111`)
 }
 
+type StringerType struct {
+	//
+}
+
+func (st *StringerType) String() string {
+	return "StringerType implements fmt.Stringer"
+}
+
 func TestEvalPointerExpressions(t *testing.T) {
 	var data = make(VarMap)
 	var s *string
 	data.Set("stringPointer", s)
-	RunJetTest(t, data, nil, "StringPointer_1", `{{ stringPointer }}`, "")
+	RunJetTest(t, data, nil, "StringPointer_1", `{{ stringPointer }}`, "<nil>")
 
 	s2 := "test"
 	data.Set("stringPointer2", &s2)
@@ -389,7 +397,30 @@ func TestEvalPointerExpressions(t *testing.T) {
 
 	var i *int
 	data.Set("intPointer", &i)
-	RunJetTest(t, data, nil, "IntPointer_i", `{{ intPointer }}`, "")
+	RunJetTest(t, data, nil, "IntPointer_i", `{{ intPointer }}`, "<nil>")
+
+	st := StringerType{}
+	data.Set("stringerType", &st) // stringerType won't be dereferenced
+	RunJetTest(t, data, nil, "StringerTypePointer", `{{ stringerType }}`, "StringerType implements fmt.Stringer")
+
+	st2 := &StringerType{}
+	data.Set("stringerType2", &st2) // stringerType2 will be dereferenced just once
+	RunJetTest(t, data, nil, "StringerTypePointer2", `{{ stringerType2 }}`, "StringerType implements fmt.Stringer")
+
+	u := User{
+		Name:  "Pablo Escobbar",
+		Email: "pablo.escobar@cartel.mx",
+	}
+	data.Set("user", &u) // user will be dereferenced once
+	RunJetTest(t, data, nil, "UserPointer", `{{ user }}`, "{Pablo Escobbar pablo.escobar@cartel.mx}")
+
+	u1 := &u
+	data.Set("user1", &u1) // user1 will be dereferenced twice
+	RunJetTest(t, data, nil, "UserPointer1", `{{ user1 }}`, "{Pablo Escobbar pablo.escobar@cartel.mx}")
+
+	u2 := &u1
+	data.Set("user2", &u2) // user2 will be dereferenced only twice
+	RunJetTest(t, data, nil, "UserPointer2", `{{ user2 }}`, "&{Pablo Escobbar pablo.escobar@cartel.mx}")
 }
 
 func TestEvalPointerLimitNumberOfDereferences(t *testing.T) {
@@ -397,15 +428,56 @@ func TestEvalPointerLimitNumberOfDereferences(t *testing.T) {
 
 	var i *int
 	data.Set("intPointer", &i)
-	RunJetTest(t, data, nil, "IntPointer_i", `{{ intPointer }}`, "")
+	RunJetTest(t, data, nil, "IntPointer_i", `{{ intPointer }}`, "<nil>")
 
 	j := &i
 	data.Set("intPointer", &j)
-	RunJetTest(t, data, nil, "IntPointer_j", `{{ intPointer }}`, "")
+	RunJetTest(t, data, nil, "IntPointer_j", `{{ intPointer }}`, "<nil>")
 
-	k := &j
-	data.Set("intPointer", &k)
-	RunJetTest(t, data, nil, "IntPointer_1", `{{ intPointer }}`, "<nil>")
+	// k := &j
+	// data.Set("intPointer", &k)
+	// RunJetTest(t, data, nil, "IntPointer_1", `{{ intPointer }}`, "<nil>")
+}
+
+type Apple struct {
+	Flavor string
+}
+
+func (a *Apple) GetFlavor() string {
+	return a.Flavor
+}
+
+func (a *Apple) GetFlavorPtr() *string {
+	return &a.Flavor
+}
+
+func TestApple(t *testing.T) {
+
+	apples := map[string]*Apple{
+		"honeycrisp": {
+			Flavor: "crisp",
+		},
+		"red-delicious": {
+			Flavor: "poor",
+		},
+		"granny-smith": {
+			Flavor: "tart",
+		},
+	}
+
+	var data = make(VarMap)
+	data.SetFunc("GetAppleByName", func(a Arguments) reflect.Value {
+		name := a.Get(0).String()
+		return reflect.ValueOf(apples[name])
+	})
+
+	data.SetFunc("TellFlavor", func(a Arguments) reflect.Value {
+		apple := a.Get(0).Interface().(*Apple)
+		flav := apple.GetFlavor()
+		return reflect.ValueOf(flav)
+	})
+
+	RunJetTest(t, data, nil, "LookUpApple", `{{apple := GetAppleByName("honeycrisp")}}{{TellFlavor(apple)}}`, "crisp")
 }
 
 func TestEvalStructFieldPointerExpressions(t *testing.T) {

--- a/eval_test.go
+++ b/eval_test.go
@@ -433,10 +433,6 @@ func TestEvalPointerLimitNumberOfDereferences(t *testing.T) {
 	j := &i
 	data.Set("intPointer", &j)
 	RunJetTest(t, data, nil, "IntPointer_j", `{{ intPointer }}`, "<nil>")
-
-	// k := &j
-	// data.Set("intPointer", &k)
-	// RunJetTest(t, data, nil, "IntPointer_1", `{{ intPointer }}`, "<nil>")
 }
 
 type Apple struct {

--- a/func.go
+++ b/func.go
@@ -26,6 +26,11 @@ type Arguments struct {
 	argVal  []reflect.Value
 }
 
+// IsSet checks whether an argument is set or not. It behaves like the build-in isset function.
+func (a *Arguments) IsSet(argumentIndex int) bool {
+	return a.runtime.isSet(a.argExpr[argumentIndex])
+}
+
 // Get gets an argument by index.
 func (a *Arguments) Get(argumentIndex int) reflect.Value {
 	if argumentIndex < len(a.argVal) {

--- a/go.mod
+++ b/go.mod
@@ -4,4 +4,4 @@ go 1.12
 
 require github.com/CloudyKit/fastprinter v0.0.0-20170127035650-74b38d55f37a
 
-replace github.com/CloudyKit/fastprinter => github.com/annismckenzie/fastprinter v0.0.0-20180721091441-b9f2fb63ce50 // PR 1 at https://github.com/CloudyKit/fastprinter/pull/1
+replace github.com/CloudyKit/fastprinter => github.com/sauerbraten/fastprinter v0.0.0-20200109142611-26ce1e0b7924

--- a/go.mod
+++ b/go.mod
@@ -2,6 +2,4 @@ module github.com/CloudyKit/jet/v3
 
 go 1.12
 
-require github.com/CloudyKit/fastprinter v0.0.0-20170127035650-74b38d55f37a
-
-replace github.com/CloudyKit/fastprinter => github.com/sauerbraten/fastprinter v0.0.0-20200109142611-26ce1e0b7924
+require github.com/CloudyKit/fastprinter v0.0.0-20200109182630-33d98a066a53

--- a/node.go
+++ b/node.go
@@ -61,7 +61,7 @@ func (node *NodeBase) error(err error) {
 }
 
 func (node *NodeBase) errorf(format string, v ...interface{}) {
-	panic(fmt.Errorf("Jet Runtime Error(%q:%d): %s", node.TemplateName, node.Line, fmt.Sprintf(format, v...)))
+	panic(fmt.Errorf("jet runtime error @ %q:%d: %s", node.TemplateName, node.Line, fmt.Sprintf(format, v...)))
 }
 
 // Type returns itself and provides an easy default implementation

--- a/node.go
+++ b/node.go
@@ -61,7 +61,7 @@ func (node *NodeBase) error(err error) {
 }
 
 func (node *NodeBase) errorf(format string, v ...interface{}) {
-	panic(fmt.Errorf("jet runtime error @ %q:%d: %s", node.TemplateName, node.Line, fmt.Sprintf(format, v...)))
+	panic(fmt.Errorf("Jet Runtime Error (%q:%d): %s", node.TemplateName, node.Line, fmt.Sprintf(format, v...)))
 }
 
 // Type returns itself and provides an easy default implementation

